core/src/main/scala/org/apache/spark/status/storeTypes.scala

```diff
@@ -76,6 +109,29 @@ private[spark] class JobDataWrapper(
 
   @JsonIgnore @KVIndex("completionTime")
   private def completionTime: Long = info.completionTime.map(_.getTime).getOrElse(-1L)
+
+  def toLiveJob: LiveJob = {
```

Comment:
`completedIndices` and `completedStages` are not assigned, and they're not computed from `AppStatusListener.recoverLiveEntities`. Is it missing, or valid reason to do so, or technical issue?

Reply:
While converting `LiveJob` to `JobDataWrapper`,  only `completedIndices.size`/`completedStages.size` are written into KVStore. So, it's impossible to recover the detail `completedIndices`/`completedStages`. So, as a compromises, we only recover the number in this case. 

And this brings a problem that the recovered live job wouldn't be 100% the same as the previous one. But, this can be acceptable when you look at their(`completedIndices`, `completedStages`) usages in `AppStatusListener`, as it does not mistake the final result(that is number).

Reply:
you can get multiple taskend events for the same `taskInfo.index` however, so you actually need those indices

Reply:
Yeah, I can see how that would happen. But I think there would be only one successful task out of those end tasks, right ? And we only update `completedIndices` when we receive a successful end task:

https://github.com/apache/spark/blob/d2f21b019909e66bf49ad764b851b4a65c2438f8/core/src/main/scala/org/apache/spark/status/AppStatusListener.scala#L579-L588

https://github.com/apache/spark/blob/d2f21b019909e66bf49ad764b851b4a65c2438f8/core/src/main/scala/org/apache/spark/status/AppStatusListener.scala#L621-L623

WDYT ? @squito

Reply:
two speculative tasks can both complete successfully.  And the scheduler intentionally posts TaskEnd events for both of them:

https://github.com/apache/spark/blob/6390f02f9fba059ec5d089a68c8d758aca35c9cd/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1369-L1375

Reply:
Oh, I see. Then, this could really be a problem which needs to be fixed.

Reply:
This is just a brainstorm, and not something I'm sure is the right approach at all yet --

we could also consider just letting some details like this be lossy.  But if we were go down that road, I'd like some mechanism to ensure that we only accepted lost info where it "probably" didn't matter.  Eg. we'd save metadata in such a way that the SHS would know to replay currently running jobs from the event log, even though some of that info had already been parsed and stored in the snapshot.  But say there was some really late speculative task completion, from a job that had finished long ago -- that may not be represented at full-fidelity.

I feel like this would cover streaming pretty well.  It would not work as well for job-server style deployments ... but, really, I can't think of anything which covers that very well.

Reply:
I wrote down my thoughts on this idea a little bit more here (I think it might be useful to collect the main ideas discussed somewhere as the discussion in comments is getting long and scattered): https://docs.google.com/document/d/1LbPflpEyrh9jDunwmDnrTjtMybAXcGimy36AWcUQsq8/edit?usp=sharing

Reply:
Well, a more simple and direct way here maybe is just to write `completedIndices` and `completedStages` out to the KVStore, too. Then, we could simply restore it later, though this would require more memory than before.

Reply:
where would you store it?  those could be sets of size 100K, I would want to make sure we aren't writing it out on every task complete.

(but to the more general point -- yes, we can go for both solutions.  Store more info when possible, but perhaps also accept some loss of info.)

Reply:
We'd write them out whenever a live job needs to be updated into KVStore with current updated logic. As it doesn't write out on every task complete but would update when exceeds configured `liveUpdatePeriodNs` or some "last" behavior happens.

But as you mentioned, this could require much memory and pay pressure on snapshotting. 

But can this info be lossy ? User may find that a Job consists of X tasks would have X + m finished tasks where extra m tasks are those duplicate completed indices with speculative enabled. 


We may use some known and accurate infos, e.g `numTasks`, `activeTasks`, `completedTasks`, `failedTasks`, to calculate the correct `completedIndices.size / completedStages.size` when a job finished.

Reply:
Any thoughts here ? @HeartSaVioR

Reply:
I'm not sure we are OK with lossy information - we shouldn't let developer access information to do something which can be lost after restoring, but how? Leaving comments? Whenever we modify AppStatusListener to add some feature, we may have to struggle with both things - compatibility with old snapshot of state, and uncertain of lossy information if we cannot control well.

Ideally we would have to restore state of listener as same as it was; if we have to snapshot the state of listener in driver, memory consumption and performance would really matter as it affects live application. For SHS, we will have less pressure on this, and actually memory consumption would less matter as in fact we are not tied to store the state of listener in "KVStore". It could be another file along with snapshot file on KVStore.

Reply:
I haven't fully followed the discussion here, not yet thought thoroughly about all the implications of the snapshotting approaches being discussed. But I just wanted to point out that the problems being talked about here (snapshotting a potentially large amount of data) are ridiculously worse on the SQL side.

Because of the way SQL metrics are calculated you need to have all data points, meaning that a snapshot taken in the middle of a stage of a SQL query can have a humongous amount of data.

The 100k task stage here could be encoded with 25k characters if you encode the bit set as hex characters (4 bits per char). That's not so bad. But on the SQL side, if you have let's say half of the tasks finished, you have to encode 50k longs * number of metrics being tracked by the stage, which is a lot more data to encode in json.

Reply:
So, will it be acceptable if we only consider snapshotting in the SHS ?

Reply:
well, the concerns about putting excess work on the driver are certainly not as important on the SHS.  Performance still matters, of course, but there maybe more options to speed things up (eg., running multiple processes to create snapshots in parallel).  And we're still concerned about having a small snapshot file (if its too big, it won't be fast to load, of course).  We dont' know for sure what that file size will be till we actually build it ... but my gut is that we can store all the required info in the snapshot file and keep things reasonably sized.